### ISSUE TITLE
Subsequent variable arithmetics / interpolation when count is zero

### ISSUE THREAD

**Issue by Artiax** [hashicorp/terraform#15993]
### Terraform Version
0.9.11

### Terraform Configuration Files
```hcl
resource "aws_route" "peer_route" {
  count                       = "${length(data.aws_vpc_peering_connection.peer_connection.*.id) > 0 ? (length(var.peer_environment_groups) * length(var.route_table_ids)) : 0}"

  route_table_id              = "${element(var.route_table_ids, ceil(count.index / length(var.peer_environment_groups)))}"
```

The above would fail with the following error (as var.peer_environment_groups in route_table_id is zero):
```hcl
module.routes_pr.aws_route.peer_route: __builtin_IntMath: divide by zero in:

${element(var.route_table_ids, ceil(count.index / length(var.peer_environment_groups)))}

```

I've got a workaround for it, however it would be nice if I didn't have to do it:
```hcl
route_table_id              = "${element(var.route_table_ids, ceil(count.index / (length(var.peer_environment_groups) > 0 ? length(var.peer_environment_groups) : 1)))}"
```

Explanation: Dividing by 1 rather than 0 when peer_environment_groups length is zero, this won't actually do anything but unblocks the plan and doesn't produce the error anymore.

### Expected Behavior
Ignore the subsequent variables/function calls when count is zero.

### Actual Behavior
Terraform proceeded to interpolate the variables and carry out the arithmetics when the count of the resource is zero.

**Comment by jcrsilva** [comment]
Also affected by this. If count==0 the other variables shouldn't be interpolated.

**Comment by farvour** [comment]
I am also impacted by this scenario.

**Comment by yruss972** [comment]
+1

**Comment by apparentlymart** [comment]
Posting +1 comments doesn't do anything except create noise for the people following this issue. To influence prioritization, please add a :+1: reaction to the original comment.

---

With this said: Terraform's behavior here is going to change slightly in the next major release to improve the situation here, but not in exactly the way described. It is intentional that Terraform validates the configuration within a `resource` block even when `count = 0` because the purpose of validation is to check for _static_ (that is, regardless of particular input variable values) problems, similar to how a type checker in a statically-typed programming language will report a problem even if it's in a codepath you aren't calling. The goal is to give you early notice of problems that might otherwise be hidden from you until some later point when you set `count` to a value greater than zero, where the person encountering the problem might be someone else entirely.

I understand that there are some situations where this behavior is inconvenient, because as the author you know that in practice a certain problem only occurs with `count = 0`. To make this more convenient in many cases while still retaining the goals of the validation pass, the new compromise is as follows:

During validation, Terraform will check the expressions under the assumption that `count.index` is a number _but_ it will not assume any particular value for that number. This situation is similar to what happens if you interpolate an attribute of a resource that won't be known until after apply (currently displayed as `<computed>` in the UI in 0.11): Terraform will be able to ensure that you are using it in a way that makes sense for a number to be used but won't raise any problems that depend on the particular value of that number.

Taking the example from the original comment here:

```hcl
route_table_id = "${element(var.route_table_ids, ceil(count.index / length(var.peer_environment_groups)))}"
```

Terraform will evaluate the argument to that `ceil` call as `<unknown number> / 0`, which is type-safe. Fortunately, the unknown value short-circuit rules in the checker will cause Terraform to skip trying to _evaluate_ this expression, because an unknown value divided by anything is another unknown value.

This should therefore pass validation and then not be considered again by later passes due to the zero count.

However (for completeness) if we were to instead write something like this:

```hcl
route_table_id = count.index - "not a number"
```

...the validation pass would still flag this as an error, even if `count = 0`, since the language knows that `"not a number"` can't possibly be a number and thus can't possibly be subtracted from a number, regardless of what value `count.index` eventually has.

It's likely that there will still be some cases where the static checker will flag things overly-conservatively. That's a consequence the safety/flexibility tradeoff we're trying to make here, but since the next major release is also planned to fix #15605 it should be possible to give the validator additional hints by placing the troublesome expression in a conditional clause that isn't evaluated when `count = 0`.

**Comment by apparentlymart** [comment]
Hi all!

The changes I mentioned in my previous comment were included in Terraform v0.12.0, which has now been released.
